### DISCUSSION TITLE
Is it possible to improve performace when listing branches/references?

### DISCUSSION THREAD

**Discussion by dschrempf** [magit/magit#5164] [Support] [answered]
This must have come up at some point but I didn't find any other related questions.

The speed of Magit is sometimes problematic when managing large repositories. In these cases, I manage to remove most of the lags with the following configuration:

    ((nil
      . ((magit-revision-insert-related-refs . nil)))
     (magit-status-mode
      . ((eval . (magit-disable-section-inserter 'magit-insert-tags-header))
         (eval . (magit-disable-section-inserter 'magit-insert-recent-commits))
         (eval . (magit-disable-section-inserter 'magit-insert-unpushed-to-pushremote))
         (eval . (magit-disable-section-inserter 'magit-insert-unpushed-to-upstream-or-recent))
         (eval . (magit-disable-section-inserter 'magit-insert-unpulled-from-pushremote))
         (eval . (magit-disable-section-inserter 'magit-insert-unpulled-from-upstream)))))

However, I do not manage to get rid of a multi-second freeze when "listing branches/references". For example, when checking out a new branch (`b b`), or rebasing on another branch. Then, Emacs freezes various seconds before listing all available branches/references (thousands in my case). Is it possible to tell Magit to NOT list those branches (or cache them, and I tell it when to refresh the cache?).

I tried tinkering around with `magit-refs-sections-hook` (e.g., `(remove-hook 'magit-refs-sections-hook 'magit-insert-tags)`), but that didn't help either.

Thank you for working on Magit!

**Comment by tarsius** [comment, marked as answer]
Caching is of course always an option, but comes with extra complexity and the issue of cache invalidation. Another option would be to only fetch the refs that are actually of interest. Note that you can specify "negative refspecs", see git-fetch(1). Limiting the number of refs seems preferable to me, since, beside making the code go brrr, the removal of clutter also makes it easier for the human to find what they are looking for.

**Reply by dschrempf** [reply]
Thank you! I didn't know about `refspec`s and they offer a much better experience than I had hoped for!

**Reply by tarsius** [reply]
Great :grinning:
